File: backend/apps/speech/tts/tts_driver.py
from abc import ABC, abstractmethod
import logging
import os
from .minimax_tts import MinimaxTTS, minimax_voices, EMOTIONS

logger = logging.getLogger(__name__)


class BaseTTS(ABC):
    '''合成语音统一抽象类'''

    @abstractmethod
    def synthesis(self, text: str, voice_id: str, **kwargs) -> str:
        '''合成语音'''
        pass

    @abstractmethod
    def get_voices(self, limit=30) -> list[dict[str, str]]:
        '''获取声音列表'''
        pass


class MinimaxTTSWrapper(BaseTTS):
    '''Minimax TTS包装类'''
    client: MinimaxTTS

    def __init__(self):
        self.client = MinimaxTTS()

    def synthesis(self, text: str, voice_id: str, **kwargs) -> str:
        emotion = kwargs.get("emotion", "neutral")
        stream = kwargs.get("stream", False)
        return self.client.create_audio(text=text, voice_id=voice_id, emotion=emotion, stream=stream)

    def get_voices(self, limit=30) -> list[dict[str, str]]:
        return minimax_voices[:limit]


class TTSDriver:
    '''TTS驱动类'''

    def __init__(self):
        self.minimax_tts = MinimaxTTSWrapper()
        self.tts_map = {
            "minimax": self.minimax_tts
        }

    def synthesis(self, text: str, tts_type: str, voice_id: str, **kwargs) -> str:
        logger.info(f"Synthesis text={text}, tts_type={tts_type}, voice_id={voice_id}, kwargs={kwargs}")
        
        try:
            # 检查TTS类型是否存在
            if tts_type not in self.tts_map:
                logger.warning(f"不支持的TTS类型: {tts_type}，强制使用minimax")
                tts_type = "minimax"
                
            tts = self.tts_map[tts_type]
            
            # 调用对应的TTS合成
            result = tts.synthesis(text, voice_id, **kwargs)
            
            # 记录结果类型以便调试
            if isinstance(result, bytes):
                logger.info(f"Synthesis result type: bytes, size: {len(result)} bytes")
            else:
                logger.info(f"Synthesis result type: {type(result)}")
                
            return result
        except Exception as e:
            logger.error(f"语音合成错误: {str(e)}")
            return None

    def get_voices(self, tts_type: str = None) -> dict[str, list[dict[str, str]]]:
        try:
            result = {}
            
            if tts_type is None or tts_type not in self.tts_map:
                # 获取所有TTS类型的语音 (实际只有minimax)
                for key, tts in self.tts_map.items():
                    result[key] = tts.get_voices()
            else:
                # 获取指定TTS类型的语音
                result[tts_type] = self.tts_map[tts_type].get_voices()
                
            return result
        except Exception as e:
            logger.error(f"获取语音列表错误: {str(e)}")
            return {}
    
    def get_emotions(self) -> list[str]:
        """获取支持的情绪列表"""
        return EMOTIONS

File: frontend/src/lib/tts/ttsDriver.ts
import { MinimaxTTS, minimaxVoices, EMOTIONS } from './minimaxTTS';
import { synthesizeVoice } from '@/features/koeiromap/koeiromap';
import fs from 'fs';
import path from 'path';

/**
 * TTS配置
 */
interface TTSConfig {
  ttsType: string;
  minimax?: {
    apiKey: string;
    groupId: string;
    model?: string;
  };
}

/**
 * 基础TTS接口
 */
export interface BaseTTS {
  /**
   * 生成语音
   * @param text 文本
   * @param voiceId 声音ID
   * @param options 选项
   */
  synthesize(text: string, voiceId: string, options?: any): Promise<any>;
  
  /**
   * 获取声音列表
   */
  getVoices(): any[];
}

/**
 * Koeiromap TTS实现
 */
export class KoeiromapTTS implements BaseTTS {
  async synthesize(text: string, voiceId: string, options: any = {}): Promise<any> {
    const speakerX = options.speakerX || 1.32;
    const speakerY = options.speakerY || 1.88;
    const style = options.style || 'talk';
    
    const voice = await synthesizeVoice(text, speakerX, speakerY, style);
    return voice;
  }
  
  getVoices(): any[] {
    return [{ id: 'default', name: '默认语音' }];
  }
}

/**
 * Minimax TTS包装类
 */
export class MinimaxTTSWrapper implements BaseTTS {
  private client: MinimaxTTS;
  
  constructor(config: { apiKey: string, groupId: string, model?: string }) {
    this.client = new MinimaxTTS(config);
  }
  
  async synthesize(text: string, voiceId: string, options: any = {}): Promise<any> {
    const emotion = options.emotion || 'neutral';
    const stream = options.stream || false;
    
    return await this.client.createAudio(text, voiceId, emotion, stream);
  }
  
  getVoices(): any[] {
    return this.client.getVoices();
  }
}

/**
 * TTS驱动类
 */
export class TTSDriver {
  private ttsMap: Record<string, BaseTTS> = {};
  
  constructor() {
    // 读取配置
    const config = this.getConfig();
    
    // 初始化Koeiromap TTS
    this.ttsMap['koeiromap'] = new KoeiromapTTS();
    
    // 初始化Minimax TTS (如果配置存在)
    if (config.minimax?.apiKey && config.minimax?.groupId) {
      this.ttsMap['minimax'] = new MinimaxTTSWrapper({
        apiKey: config.minimax.apiKey,
        groupId: config.minimax.groupId,
        model: config.minimax.model
      });
    }
  }
  
  /**
   * 读取配置
   */
  private getConfig(): TTSConfig {
    try {
      const configPath = path.join(process.cwd(), 'data', 'config.json');
      const configData = fs.readFileSync(configPath, 'utf8');
      const config = JSON.parse(configData);
      
      return config.ttsConfig || { ttsType: 'koeiromap' };
    } catch (error) {
      console.error('读取TTS配置文件失败:', error);
      return { ttsType: 'koeiromap' };
    }
  }
  
  /**
   * 合成语音
   * @param text 文本
   * @param ttsType TTS类型
   * @param voiceId 声音ID
   * @param options 选项
   */
  async synthesize(text: string, ttsType: string, voiceId: string, options: any = {}): Promise<any> {
    console.log(`Synthesis text=${text.substring(0, 30)}..., tts_type=${ttsType}, voice_id=${voiceId}`);
    
    try {
      // 检查TTS类型是否存在
      if (!this.ttsMap[ttsType]) {
        console.warn(`不支持的TTS类型: ${ttsType}，强制使用koeiromap`);
        ttsType = 'koeiromap';
      }
      
      const tts = this.ttsMap[ttsType];
      
      // 调用对应的TTS合成
      const result = await tts.synthesize(text, voiceId, options);
      
      return result;
    } catch (error) {
      console.error('语音合成错误:', error);
      throw error;
    }
  }
  
  /**
   * 获取声音列表
   * @param ttsType TTS类型
   */
  getVoices(ttsType?: string): Record<string, any[]> {
    try {
      const result: Record<string, any[]> = {};
      
      if (!ttsType) {
        // 获取所有TTS类型的语音
        for (const [key, tts] of Object.entries(this.ttsMap)) {
          result[key] = tts.getVoices();
        }
      } else if (this.ttsMap[ttsType]) {
        // 获取指定TTS类型的语音
        result[ttsType] = this.ttsMap[ttsType].getVoices();
      }
      
      return result;
    } catch (error) {
      console.error('获取语音列表错误:', error);
      return {};
    }
  }
  
  /**
   * 获取情绪列表
   */
  getEmotions(): string[] {
    return EMOTIONS;
  }
} 